pycloudlib/ibm_softlayer/cloud.py

```diff
@@ -0,0 +1,492 @@
+            notes="Added by pycloudlib",
+        )
+        key_id = new_ssh_key["id"]
+        self._log.debug("Created SSH key with id: %s", key_id)
+        self.created_keys.append(key_id)
+        return key_id
+
+    def delete_key(self, name: str):
+        """Delete SSH key by name."""
+        target_key = None
+        for key in self._ssh_key_manager.list_keys():
+            if key["label"] == name:
+                target_key = key
+                break
+        if target_key is None:
+            raise IBMSoftlayerException(f"Key with name {name} not found")
+        self._log.debug("Deleting SSH key: %s", name)
+        self._ssh_key_manager.delete_key(target_key["id"])
+
+    def create_default_security_groups(self) -> tuple[int, int]:
+        """
+        Create default security groups.
+
+        To make this extensible for all users, security groups and rules are
+        created on the fly. A unique security group is created for each
+        instance so that it can be torn down later without affecting other
+        instances. The security group is named after the instance tag. The
+        public security group allows inbound ssh traffic and all outbound
+        traffic. The private security group allows all inbound and outbound
+        traffic.
+
+        Returns:
+            tuple[int, int]: The IDs of the created security groups.
+        """
+        self._log.info("Creating default security groups")
+        # create security groups
+        public_security_group_id = self._create_security_group(
+            self.tag + "-public-security-group",
+            "Allows ssh inbound and all outbound traffic",
+        )
+        private_security_group_id = self._create_security_group(
+            self.tag + "-private-security-group",
+            "Allows all inbound and outbound traffic",
+        )
+        # allow inbound ssh for public security group
+        self._add_rules_to_security_group(
+            public_security_group_id,
+            ["ingress"],
+            port=22,
+            protocol="tcp",
+            ipv6=True,
+        )
+        # allow all outbound traffic for public security group
+        self._add_rules_to_security_group(
+            public_security_group_id,
+            ["egress"],
+            port=None,
+            protocol=None,
+            ipv6=True,
+        )
+        # allow all inbound and outbound traffic for private security group
+        self._add_rules_to_security_group(
+            private_security_group_id,
+            ["ingress", "egress"],
+            port=None,
+            protocol=None,
+            ipv6=True,
+        )
+        self._log.debug("Added rules to security groups.")
+        return public_security_group_id, private_security_group_id
+
+    def _create_security_group(self, name: str, description: str) -> int:
+        """
+        Create a security group.
+
+        Args:
+            name (str): The name of the security group.
```

Comment:
As the types are already documented as type hints, could we remove them from the docstrings? It feels repetitive and very likely to end up out of sync.

If we do so, please make it in all the docstrings.

Reply:
yeah good call. will do!

Reply:
This is still unfixed in some places.